### PR TITLE
aws-azure-login: Link Puppeteer to Chromium

### LINKED ISSUE #272544: `aws-azure-login` can't find browser
### Describe the bug

`aws-azure-login` consistently fails on 23.11.

### Steps To Reproduce
1. Upgrade to NixOS 23.11
1. Run `nix-shell -p aws-azure-login --run 'aws-azure-login --profile=some-profile'`

### Expected behavior

This should work like [`nodePackages.aws-azure-login` on 23.05](https://search.nixos.org/packages?channel=23.05&from=0&size=50&sort=relevance&type=packages&query=aws-azure-login).

### Additional context

Debug output:

```
❯ DEBUG=aws-azure-login aws-azure-login --profile=my-profile
  aws-azure-login Getting config for profile 'my-profile' in section 'profile my-profile' +0ms
  aws-azure-login Loading 'config' file at '/home/username/.aws/config' +1ms
  aws-azure-login Parsing data +1ms
  aws-azure-login Environment +0ms
  aws-azure-login { azure_default_password: 'xxxxxxxxxx' } +0ms
Logging in with profile 'my-profile'...
Using AWS SAML endpoint https://signin.aws.amazon.com/saml
  aws-azure-login Generating UUID for SAML request +2ms
  aws-azure-login Generated SAML request
        <samlp:AuthnRequest xmlns="urn:oasis:names:tc:SAML:2.0:metadata" ID="id2893d26d-0ab2-4a89-bd0a-4754e40fc327" Version="2.0" IssueInstant="2023-12-06T21:37:03.101Z" IsPassive="false" AssertionConsumerServiceURL="https://signin.aws.amazon.com/saml" xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol">
            <Issuer xmlns="urn:oasis:names:tc:SAML:2.0:assertion">1050e591-a33e-4602-9294-0cf5350f8c4d</Issuer>
            <samlp:NameIDPolicy Format="urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"></samlp:NameIDPolicy>
        </samlp:AuthnRequest>
         +0ms
  aws-azure-login Deflating SAML +0ms
  aws-azure-login Encoding SAML in base64 +3ms
  aws-azure-login Created login URL https://login.microsoftonline.com/2134e961-7e38-4c34-a22b-10da5466b725/saml2?SAMLRequest=… +0ms
  aws-azure-login Loading login page in Chrome +0ms
Error: Could not find expected browser (chrome) locally. Run `npm install` to download the correct Chromium revision (982053).
    at ChromeLauncher.launch (/nix/store/mxglfzfd0y7wkl2scx3q2slf0ngdlqk2-aws-azure-login-3.6.1/lib/node_modules/aws-azure-login/node_modules/puppeteer/lib/cjs/puppeteer/node/Launcher.js:105:23)
    at PuppeteerNode.launch (/nix/store/mxglfzfd0y7wkl2scx3q2slf0ngdlqk2-aws-azure-login-3.6.1/lib/node_modules/aws-azure-login/node_modules/puppeteer/lib/cjs/puppeteer/node/Puppeteer.js:125:31)
    at Object._performLoginAsync (/nix/store/mxglfzfd0y7wkl2scx3q2slf0ngdlqk2-aws-azure-login-3.6.1/lib/node_modules/aws-azure-login/lib/login.js:518:49)
    at async Object.loginAsync (/nix/store/mxglfzfd0y7wkl2scx3q2slf0ngdlqk2-aws-azure-login-3.6.1/lib/node_modules/aws-azure-login/lib/login.js:377:30)
```

Installing `pkgs.chromium` via `users.users.<name>.packages` or via `programs.chromium.enable` does *not* fix this.

[`nodePackages.aws-azure-login` had a lot more dependencies on 23.05](https://raw.githubusercontent.com/NixOS/nixpkgs/nixos-23.05/pkgs/development/node-packages/node-packages.nix), but no mention of Chromium itself.

It looks like the error message is from _[puppeteer](https://stackoverflow.com/q/68051648/96588)._ [Troubleshooting reference](https://github.com/puppeteer/puppeteer/blob/06cacfa9c1d5b185ce41a337abf94cfa8726b4f2/docs/troubleshooting.md#could-not-find-expected-browser-locally).

### Notify maintainers

@yurrriq

### Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.64, NixOS, 23.11 (Tapir), 23.11.1209.933d7dc15509`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos-23.11, nixos-unstable"`
 - channels(username): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### Failed workaround

(I'll be tweaking this in case I come up with a working solution.)

1. Go to a new directory, for example `aws-azure-login`
1. Create a `package.json` containing the following:

   ```
   {
     "name": "aws-azure-login",
     "dependencies": {
       "aws-azure-login": "*"
     }
   }
   ```
1. Run `node2nix --nodejs-18`
1. Get the built package like this:

   ```
   awsAzureLoginNodePackages = import ./aws-azure-login/default.nix {
     inherit pkgs;
   };
   awsAzureLogin =
     (awsAzureLoginNodePackages
       // {
         puppeteer = awsAzureLoginNodePackages.puppeteer.override {
           env.PUPPETEER_SKIP_DOWNLOAD = true;
         };
       })
     .aws-azure-login;
   ```

Unfortunately, the `PUPPETEER_SKIP_DOWNLOAD` override doesn't work as expected:

```
ERROR: Failed to set up Chromium r982053! Set "PUPPETEER_SKIP_DOWNLOAD" env variable to skip download.

I'm not sure how to proceed.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by l0b0**
## Description of changes

Closes #272544, avoiding the following runtime error:

> Error: Could not find expected browser (chrome) locally. Run `npm install` to download the correct Chromium revision (982053).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/aw/aws-azure-login/package.nix
Comment (l0b0) [reply pkgs/by-name/aw/aws-azure-login/package.nix:10]: Oh, that was just the auto-formatter.

File: pkgs/by-name/aw/aws-azure-login/package.nix
Comment (yurrriq) [inline pkgs/by-name/aw/aws-azure-login/package.nix:98]: Thanks for this!

File: pkgs/by-name/aw/aws-azure-login/package.nix
Comment (imincik) [inline pkgs/by-name/aw/aws-azure-login/package.nix:3]: remove unused `callPackage`

File: pkgs/by-name/aw/aws-azure-login/package.nix
Comment (imincik) [inline pkgs/by-name/aw/aws-azure-login/package.nix:9]: Don't import all `pkgs` for using `pkgs.runCommand`. Just import `runCommand` instead .

File: pkgs/by-name/aw/aws-azure-login/package.nix
Comment (imincik) [inline pkgs/by-name/aw/aws-azure-login/package.nix:72]: Use `runCommand` instead of `pkgs.runCommand` as suggested above.

**Review by imincik — CHANGES_REQUESTED** [review]
Thanks for fixing the package. Fix looks good to me but I'll made some comments and requested few changes. After that, I am good to merge this PR.

Here is the example of using `finalAttrs` function plus some other changes I requested:

```diff
diff --git a/pkgs/by-name/aw/aws-azure-login/package.nix b/pkgs/by-name/aw/aws-azure-login/package.nix
index e6eaa7fde372..164be566ee2f 100644
--- a/pkgs/by-name/aw/aws-azure-login/package.nix
+++ b/pkgs/by-name/aw/aws-azure-login/package.nix
@@ -1,28 +1,28 @@
 { lib
+, runCommand
 , stdenv
-, callPackage
+
 , chromium
 , fetchFromGitHub
 , fetchYarnDeps
 , makeWrapper
 , nodejs
-, pkgs
 , prefetch-yarn-deps
 , yarn
 }:
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
   pname = "aws-azure-login";
   version = "3.6.1";

   src = fetchFromGitHub {
     owner = "aws-azure-login";
     repo = "aws-azure-login";
-    rev = "v${version}";
+    rev = "v${finalAttrs.version}";
     hash = "sha256-PvPnqaKD98h3dCjEOwF+Uc86xCJzn2b9XNHHn13h/2Y=";
   };

   offlineCache = fetchYarnDeps {
-    yarnLock = "${src}/yarn.lock";
+    yarnLock = "${finalAttrs.src}/yarn.lock";
     hash = "sha256-SXQPRzF6b1FJl5HkyXNm3kGoNSDXux+0RYXBX93mOts=";
   };

@@ -69,7 +69,7 @@ stdenv.mkDerivation rec {
   '';

   passthru.tests.aws-azure-login =
-    pkgs.runCommand "${pname}-tests"
+    runCommand "${finalAttrs.pname}-tests"
       {
         HOME = "/tmp/home";
       } ''
@@ -84,7 +84,7 @@ stdenv.mkDerivation rec {
       azure_default_remember_me=false
       EOF

-      ! ${lib.getExe pkgs.aws-azure-login} --profile=my-profile 2> stderr
+      ! ${lib.getExe finalAttrs.finalPackage} --profile=my-profile 2> stderr
       [[ "$(cat stderr)" == 'Unable to recognize page state! A screenshot has been dumped to aws-azure-login-unrecognized-state.png. If this problem persists, try running with --mode=gui or --mode=debug' ]]

       touch $out
@@ -98,4 +98,4 @@ stdenv.mkDerivation rec {
     maintainers = with lib.maintainers; [ l0b0 ];
     platforms = lib.platforms.all;
   };
-}
+})
```

File: pkgs/by-name/aw/aws-azure-login/package.nix
Comment (l0b0) [reply pkgs/by-name/aw/aws-azure-login/package.nix:87]: Nice, I didn't know avoiding `rec` was that easy!

File: pkgs/by-name/aw/aws-azure-login/package.nix
Comment (l0b0) [reply pkgs/by-name/aw/aws-azure-login/package.nix:2]: It's now used.

**Comment by imincik** [comment]
Result of `nixpkgs-review pr 272592` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>aws-azure-login</li>
  </ul>
</details>

**Review by imincik — APPROVED** [review]
Looks good to me now. Will merge once ofborg-eval is completed.

**Pull request merged**
